Document how to manipulate SNI with HttpClient (#36971)
dotnet/docs: docs/core/extensions/httpclient-sni.md

---
title: Customize SNI in HTTP requests
description: Learn How to control Server Name Indication TLS extension in HTTP requests.
author: rzikm
ms.author: radekzikmund
ms.date: 9/5/2023
---

# Customize SNI in HTTP requests

> When negotiating an HTTPS connection, a TLS connection needs to be established first. As part of the TLS handshake, the client sends a domain name of the server it is connecting to in one of the TLS extensions. When hosting multiple (virtual) servers on the same machine, this feature of TLS protocol allows distinguishing which of these servers clients are connecting to and configure TLS settings, such as the server certificate, accordingly.

Reviewer: ```suggestion
When negotiating an HTTPS connection, a TLS connection needs to be established first. As part of the TLS handshake, the client sends the domain name of the server it's connecting to in one of the TLS extensions. When hosting multiple (virtual) servers on the same machine, this feature of the TLS protocol allows clients to distinguish which of these servers they're connecting to and to configure TLS settings, such as the server certificate, accordingly.
```

> When making a HTTP request using `HttpClient`, the implementation automatically selects value for the SNI extension based on the URL the client is connecting to. For scenarios which require more manual control of the extension, you can use one of the following ways.

Reviewer: ```suggestion
When making an HTTP request using `HttpClient`, the implementation automatically selects a value for the server name indication (SNI) extension based on the URL the client is connecting to. For scenarios that require more manual control of the extension, you can use one of the following approaches.
```

## Host header

> Host HTTP header performs similar function as the SNI extension in TLS. It lets target server distinguish among requests for multiple host names on a single IP address. `HttpClient` will automatically fill in the Host header using the request URI. However, you can also set its value manually, and `HttpClient` will also use the new value in the SNI extension. You can use either `HttpRequestMessage.Headers.Host` or `HttpClient.DefaultRequestHeaders.Host` to achieve this effect.

Reviewer: ```suggestion
Host HTTP header performs a similar function as the SNI extension in TLS. It lets the target server distinguish among requests for multiple host names on a single IP address. `HttpClient` automatically fills in the Host header using the request URI. However, you can also set its value manually, and `HttpClient` will also use the new value in the SNI extension. You can use either `HttpRequestMessage.Headers.Host` or `HttpClient.DefaultRequestHeaders.Host` to achieve this effect.
```

```csharp
using HttpClient client = new();

client.DefaultRequestHeaders.Host = "www.microsoft.com";

using var response = await client.GetAsync("https://127.0.0.1:5001/");

System.Console.WriteLine(response);
```

> [!NOTE]

> > This method will not allow you to avoid sending SNI altogether when connecting to a URL with a hostname. If the header is set to empty string, `HttpClient` will use the hostname from the URL instead.

Reviewer: ```suggestion
> This method doesn't allow you to avoid sending SNI altogether when connecting to a URL with a hostname. If the header is set to empty string, `HttpClient` uses the hostname from the URL instead.
```

> [!NOTE]
> Customizing the Host header affects server certificate validation. By default, client will expect the server certificate to match the hostname in the Host header.

## Manual SslStream authentication via ConnectCallback

> A more complicated, but more powerful option is to use the `SocketsHttpHandler.ConnectCallback`. Since .NET 7, it is possible to return authenticated `SslStream` and thus customise how the TLS connection is being established. Inside the callback, arbitrary `SslClientAuthenticationOptions` can be used to perform client-side authentication.

Reviewer: ```suggestion
A more complicated, but also more powerful, option is to use the `SocketsHttpHandler.ConnectCallback`. Since .NET 7, it is possible to return an authenticated `SslStream` and thus customize how the TLS connection is established. Inside the callback, arbitrary `SslClientAuthenticationOptions` options can be used to perform client-side authentication.
```

```csharp
var handler = new SocketsHttpHandler
{
    ConnectCallback = async (context, cancellationToken) =>
    {
        var socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
        await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);

        var sslStream = new SslStream(new NetworkStream(socket));
        await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
        {

>             TargetHost = context.DnsEndPoint.Host,

Reviewer 2: Btw, did you check whether it's possible to skip sending SNI entirely, or do we block that somewhere else in the stack?

Reviewer 2: Nice, thanks

>         var socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
>         await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);
> 
>         var sslStream = new SslStream(new NetworkStream(socket));
>         await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
>         {
>             TargetHost = context.DnsEndPoint.Host,
> 
>         }, cancellationToken);
> 
>         return sslStream;

Reviewer 2: ```suggestion
        var socket = new Socket(SocketType.Stream, ProtocolType.Tcp) { NoDelay = true };
        try
        {
            await socket.ConnectAsync(context.DnsEndPoint, cancellationToken);

            var sslStream = new SslStream(new NetworkStream(socket, ownsSocket: true));

            // When using HTTP/2, you must also keep in mind to set options like ApplicationProtocols
            await sslStream.AuthenticateAsClientAsync(new SslClientAuthenticationOptions
            {
                TargetHost = context.DnsEndPoint.Host,

            }, cancellationToken);

            return sslStream;
        }
        catch
        {
            socket.Dispose();
            throw;
        }
```

This closer matches how we usually write `ConnectCallback` examples.
